drv/stm32h7-hash/src/lib.rs

```diff
@@ -0,0 +1,364 @@
+                self.remainder |= (data[offset] as u32) << self.nvalid;
+                self.nvalid += 8;
+                offset += 1;
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn finalize_sha256(&mut self, out: &mut [u8]) -> Result<(), HashError> {
+        #[derive(AsBytes)]
+        #[repr(C)]
+        struct Sha256 {
+            hr: [u32; SHA256_SZ / SIZEOF_U32],
+        }
+
+        match self.state {
+            State::Uninitialized => {
+                return Err(HashError::NotInitialized);
+            }
+            State::Processing => {
+                self.state = State::Finalize;
+            }
+            _ => {
+                // Trying to run finalize having written no data is an error.
+                return Err(HashError::InvalidState);
+            }
+        };
+
+        if self.nvalid > 0 {
+            // There are remainder bits that need to be written.
+            self.write_word(self.remainder, (self.nvalid / 8).into());
+            self.nvalid = 0;
+        }
+        if self.idx > 0 {
+            self.write_block(); // flush any final block
+        }
+
+        // Enable interrupt for sum calculation done.
+        self.reg
+            .imr
+            .modify(|_, w| w.dcie().set_bit().dinie().clear_bit());
+
+        if self.is_busy() {
+            while self.is_busy() {
+                hl::sleep_for(1);
+            }
+        }
+        sys_irq_control(self.interrupt, true);
+        self.reg.str.modify(|_, w| w.dcal().set_bit());
+
+        // wait for calculation to finalize and interrupt
+        loop {
+            if self.is_busy() {
+                while self.is_busy() {
+                    hl::sleep_for(1);
+                }
+            }
+            let _rm = sys_recv_closed(&mut [], self.interrupt, TaskId::KERNEL)
+                .unwrap();
+            if self.reg.sr.read().dcis().bit() {
+                break;
+            }
+            sys_irq_control(self.interrupt, true); // XXX need this again?
+            hl::sleep_for(1);
+        }
+        // DCAL is supposedly not clearable by SW.
+        // self.reg.str.modify(|_, w| w.dcal().clear_bit());
+        self.reg.imr.modify(|_, w| w.dcie().clear_bit());
+
+        // Mark as complete
+
+        // The hash is read out as words into little endian ARM world.
+        // Since the bit order needs to be maintained, read as B.E.
+        let result = Sha256 {
+            hr: [
+                u32::from_be(self.reg.hash_hr0.read().bits()),
+                u32::from_be(self.reg.hash_hr1.read().bits()),
+                u32::from_be(self.reg.hash_hr2.read().bits()),
+                u32::from_be(self.reg.hash_hr3.read().bits()),
+                u32::from_be(self.reg.hash_hr4.read().bits()),
+                u32::from_be(self.reg.hash_hr5.read().bits()),
+                u32::from_be(self.reg.hash_hr6.read().bits()),
+                u32::from_be(self.reg.hash_hr7.read().bits()),
+            ],
+        };
```

Comment:
This seems to be the only place the `Sha256` type is used, do you still need it?

Reply:
I no longer need it. I would like to iterate over the register set directly into the destination, but the Sha256 struct can go.